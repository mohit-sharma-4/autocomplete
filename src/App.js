import { useState } from 'react';
import SearchInput from './components/searchInput/SearchInput';
import Suggestions from './components/suggestions/Suggestions';
import SummaryCard from './components/summaryCard/SummaryCard';
import { countWordOccurances, debounce } from './utils/helpers';
import Data from './services/Data.json';

import './App.scss';

const App = () => {
  const [inputVal, setInputVal] = useState('');
  const [filteredData, setFilteredData] = useState([]);
  const [displayData, setDisplayData] = useState([]);

  const onSearch = debounce((searchWord) => {
    const filteredSummaries = Data.summaries.filter((value) =>
      value.summary.toLowerCase().includes(searchWord.toLowerCase())
    );

    const sortedData = sortData(filteredSummaries, searchWord);

    if (searchWord === '') {
      setFilteredData([]);
    } else {
      setFilteredData(sortedData);
    }
  }, 500);

  const handleInputChange = (e) => {
    const searchWord = e.target.value;
    setInputVal(searchWord);
    onSearch(searchWord);
  };

  const sortData = (input, word) => {
    const updatedSummaries = input.map((el) => {
      return {
        id: el.id,
        summary: el.summary,
        count: countWordOccurances(el.summary, word)
      };
    });

    return updatedSummaries.sort((a, b) => b.count - a.count);
  };

  return (
    <div className='App' data-testid='autocomplete-page'>
      <header className='app-header'>
        <h2>Autocomplete</h2>
      </header>
      <main className='app-content'>
        <div className='autocomplete-container'>
          <SearchInput
            inputVal={inputVal}
            setInputVal={setInputVal}
            onChange={handleInputChange}
          />
          {filteredData.length > 0 && (
            <Suggestions
              data={filteredData}
              setDisplayData={setDisplayData}
              setFilteredData={setFilteredData}
            />
          )}
        </div>
        <div className='card-layout-container'>
          {displayData.length > 0 &&
            displayData.map((card) => {
              const title = Data.titles[card.id];
              const author = Data.authors[card.id].author;
              const summary = card.summary;
              return (
                <SummaryCard
                  key={card.id}
                  title={title}
                  author={author}
                  summary={summary}
                />
              );
            })}
        </div>
      </main>
    </div>
  );
};

export default App;
