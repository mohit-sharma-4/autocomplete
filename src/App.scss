.App {
  height: 100vh;
  width: 100%;

  .app-header {
    height: 70px;
    width: 100%;
    background-color: #970747;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;

    h2 {
      color: #e5eaf5;
    }
  }

  .app-content {
    margin-top: 15px;
    height: 100%;

    .autocomplete-container {
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    .card-layout-container {
      padding: 0 20px;
      display: grid;
      grid-template-columns: 1fr 1fr 1fr 1fr;
      gap: 20px;
      position: absolute;
      top: 150px;
      margin-bottom: 100px;
    }
  }
}
