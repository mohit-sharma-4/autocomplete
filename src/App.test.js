/* eslint-disable testing-library/render-result-naming-convention */
/* eslint-disable testing-library/prefer-screen-queries */
import { render } from '@testing-library/react';
import App from './App';

jest.mock('./components/searchInput/SearchInput', () => () => {
  return <mock-searchinput data-testid='search-input' />;
});
jest.mock('./components/suggestions/Suggestions', () => () => {
  return <mock-suggestions data-testid='suggestions' />;
});
jest.mock('./components/summaryCard/SummaryCard', () => () => {
  return <mock-summarycard data-testid='summarycard' />;
});

const mockHandleChange = jest.fn();

describe('app', () => {
  it('should render autocomplete app on screen', () => {
    const component = render(<App />);
    const { getByTestId } = component;
    expect(getByTestId('autocomplete-page')).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it('should call onSearch when user types in input field', () => {
    const { getByTestId } = render(<App />);
    const input = getByTestId('search-input');
    expect(input).toBeDefined();
    input.addEventListener('change', mockHandleChange);
    expect(mockHandleChange).toHaveBeenCalledTimes(0);

    input.dispatchEvent(new Event('change', { bubbles: true }));
    expect(mockHandleChange).toHaveBeenCalledTimes(1);
  });
});
