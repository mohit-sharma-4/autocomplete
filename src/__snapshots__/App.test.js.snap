// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should render autocomplete app on screen 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
        data-testid="autocomplete-page"
      >
        <header
          class="app-header"
        >
          <h2>
            Autocomplete
          </h2>
        </header>
        <main
          class="app-content"
        >
          <div
            class="autocomplete-container"
          >
            <mock-searchinput
              data-testid="search-input"
            />
          </div>
          <div
            class="card-layout-container"
          />
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
      data-testid="autocomplete-page"
    >
      <header
        class="app-header"
      >
        <h2>
          Autocomplete
        </h2>
      </header>
      <main
        class="app-content"
      >
        <div
          class="autocomplete-container"
        >
          <mock-searchinput
            data-testid="search-input"
          />
        </div>
        <div
          class="card-layout-container"
        />
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
